Implement feature #5:Settlement suggestion algorithm and unit tests

ohkyounghun opened the pull request:
This pull request adds the Settlement feature (#5), which determines how debts between participants are balanced.

Changes made:
	•	Added the Settlement.suggest() method that pairs debtors and creditors using a simple greedy approach.
	•	Added a condition to skip very small balance differences (less than 0.005).
	•	Created SettlementTest.java to make sure the algorithm works correctly and handles small cases properly.

How to test:
	1.	Run mvn test in the project.
	2.	Check that all tests in SettlementTest pass.
	•	simpleTest checks a one-to-one settlement.
	•	tinyAmountTest checks that values below 0.005 are ignored.

Related issue: closes #5

gunuzello committed:
feat(settlement): implement debt-credit matching algorithm

Implemented the core settlement algorithm that matches debtors and creditors.
The method `suggest()` uses Balance.asMap() to generate a list of payments
based on remaining positive and negative balances. Debtors and creditors are
sorted and matched greedily until all balances are settled within EPS = 0.005.

ohkyounghun committed:
test: add unit tests for Settlement.suggest()

Added unit tests to verify basic and near-zero balance cases.
simpleTest checks one-to-one settlement, and tinyAmountTest ensures
balances below 0.005 are ignored correctly.

kwon-dohun committed:
Add explanations

kwon-dohun committed:
change the issue number

ohkyounghun committed:
remove .DS_Store and update .gitignore

ohkyounghun committed:
Create .gitignore

ohkyounghun committed:
Merge branch 'main' into feat/issue5-settlement

hayat711 commented:
**Group5** will be reviewing this PR [code implemented by group 6 for issue #5 ]

Kimhyewon0621 commented on src/main/java/itm/oss/splitter/Settlement.java:
```diff
@@ -1,12 +1,69 @@
+    // Split into debtors (negative) and creditors (positive)
+    List<Map.Entry<String, BigDecimal>> neg = new ArrayList<>();
+    List<Map.Entry<String, BigDecimal>> pos = new ArrayList<>();
+    BigDecimal eps = new BigDecimal("0.005");
```
I reviewed your implementation and it looks really clear structure and easy to follow!
However, I think it would be even better if a small change were made here.

Currently, the 0.005 threshold is hardcoded inside the method. Could you define it as a class-level constant.

It will be easier to reuse in other methods for rounding comparisons, 
and it would also improve maintainability in case the tolerance value ever needs to be adjusted later.

gunuzello replied:
Thanks for the suggestion!
I extracted the 0.005 tolerance to a class-level constant (EPS) and added a short comment.
This makes the threshold easier to reuse and adjust later.
(Applied to filtering, min-amount check, and “move to next person” conditions.)

Kimhyewon0621 replied:
Thanks for applying my suggestion! 
Extracting it to the EPS constant looks much cleaner and more maintainable. LGTM!

AyeongKwon commented on src/test/java/itm/oss/splitter/SettlementTest.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Test
+    void tinyAmountTest() {
```
The current name `tinyAmountTest` is a bit vague regarding the expected outcome of the test.

Following the naming convention of `should_DoSomething_WhenConditionIsMet` or `test_whenCondition_thenExpectedResult`, renaming the function would clarify the test's purpose without having to read all the function.

What do you think about one of these alternatives?

```suggestion
void test_ignoresPayments_whenBalanceIsBelowThreshold() {

mingyeonggg commented on src/main/java/itm/oss/splitter/Settlement.java:
```diff
@@ -1,12 +1,69 @@
+    BigDecimal eps = new BigDecimal("0.005");
+
+    for (Map.Entry<String, BigDecimal> e : map.entrySet()) {
+      BigDecimal v = e.getValue();
```
The variable name `v` is a bit ambiguous. Renaming it to `value` would make the code clearer.

kwon-dohun replied:
Renamed v → value as suggested. Much clearer now, thanks!

kwon-dohun replied:
Renamed v → value as suggested. Much clearer now, thanks!

mingyeonggg replied:
Great! Thanks for the update!!

AyeongKwon commented on src/test/java/itm/oss/splitter/SettlementTest.java:
```diff
@@ -0,0 +1,33 @@
+public class SettlementTest {
+
+    @Test
+    void simpleTest() {
```
The function name `simpleTest` is too simple and doesn't show the expected outcome of the test.

Renaming name following the 'test_whenCondition_thenExpectedResult' pattern would clearly state the condition and the expected outcome. This helps reviewers understand the test's purpose instantly.

I suggest renaming it to:

```suggestion
void test_singlePayment_whenOnePersonOwesAnother() {

ohkyounghun replied:
Thank you for your helpful suggestion! I agree that renaming the method makes it clearer, so I’ll apply the change.

AyeongKwon replied:
Thanks for your addressing that. It looks good to me!

AyeongKwon commented on src/main/java/itm/oss/splitter/Settlement.java:
```diff
@@ -1,12 +1,69 @@
+    // Split into debtors (negative) and creditors (positive)
+    List<Map.Entry<String, BigDecimal>> neg = new ArrayList<>();
+    List<Map.Entry<String, BigDecimal>> pos = new ArrayList<>();
+    BigDecimal eps = new BigDecimal("0.005");
+
+    for (Map.Entry<String, BigDecimal> e : map.entrySet()) {
+      BigDecimal v = e.getValue();
+      if (v == null)
+        continue;
+      if (v.compareTo(BigDecimal.ZERO) < 0 && v.abs().compareTo(eps) > 0)
+        neg.add(e);
+      else if (v.compareTo(BigDecimal.ZERO) > 0 && v.abs().compareTo(eps) > 0)
+        pos.add(e);
+    }
```
The logic inside this loop is critical for filtering valid balances and ignoring tiny amounts based on `eps`. Adding a few comments would greatly improve the clarity more for future maintainers.

I suggest adding comments to explain:
1. The purpose of skipping `null` values.
2. The role of `eps` in filtering balances before adding them to `neg` or `pos` lists.

I suggest commenting like:
```suggestion
        for (Map.Entry<String, BigDecimal> e : map.entrySet()) {
            BigDecimal v = e.getValue();
            if (v == null) continue; // Skip entries with null balance (should not happen if data is clean)
            
            // Only consider negative balances larger than epsilon (0.005) for settlement
            if (v.compareTo(BigDecimal.ZERO) < 0 && v.abs().compareTo(eps) > 0) 
                neg.add(e);
            // Only consider positive balances larger than epsilon (0.005) for settlement
            else if (v.compareTo(BigDecimal.ZERO) > 0 && v.abs().compareTo(eps) > 0) 
                pos.add(e);
        }

ohkyounghun replied:
Thank you for the detailed feedback! I agree that adding comments for these parts will make the code easier to understand. I’ll update it accordingly.

AyeongKwon replied:
Perfect!! It's much clearer now and easy to follow. Thanks for implementing the change.

Kimhyewon0621 commented on src/main/java/itm/oss/splitter/Settlement.java:
```diff
@@ -1,12 +1,69 @@
+      // Move to next person if balance is close to zero
+      if (d.getValue().abs().compareTo(eps) <= 0)
+        i++;
+      if (c.getValue().abs().compareTo(eps) <= 0)
```
I reviewed your implementation. Using BigDecimal instead of double for concurrency calculations is a good choice. 
However, I think that if you change a little in this line, your project will be more stable.

Since c.getValue() comes from the pos list which only contains positive values, the .abs() call here isn't functionally necessary, unlike on line 59 for d.getValue().abs(). So I think deleting .abs() is more explicit.

gunuzello replied:
Good point! Since c is taken from the positive list, its value never becomes negative after updates.
I removed the unnecessary .abs() and now compare c.getValue() directly with EPS.

Kimhyewon0621 replied:
LGTM!! Thank you for removing .abs(). That version looks more stable!

mingyeonggg commented on src/test/java/itm/oss/splitter/SettlementTest.java:
```diff
@@ -0,0 +1,41 @@
+public class SettlementTest {
+
+    @Test
+    void simpleTest() {
```
The current name `simpleTest` doesn't describe well about what's being tested. To improve clarity and follow the `test_whenCondition_theExpectedResult` naming convention, what about renaming it to be more descriptive? This makes the test's purpose clear just from reading its name. I suggest renaming it to `test_whenOneToOneBalance_thenSuggestsSinglePayment`.

kwon-dohun replied:
Fixed in latest commit. Thanks for the clear suggestion!

kwon-dohun replied:
Fixed in latest commit. Thanks for the clear suggestion!

mingyeonggg replied:
LGTM Thanks for applying the suggestions!

Kimhyewon0621 reviewed (commented):
I review on your implemetation. And overall, this is excellent code.

AyeongKwon reviewed (commented):
I've left a few minor suggestions on code style. Good work overall!

mingyeonggg reviewed (commented):
I've finished my review and left a few minor suggestions. Overall, this is looking good. Thanks!

kwon-dohun committed:
Deleted Table

kwon-dohun committed:
Merge branch 'main' into feat/issue5-settlement

jobregon1212 commented:
Hello guys, this PR should only modify the .java files (implementation and test files). It should not touch .gitignore , README, or any other file. So please fix that. Remember to synchronize with upstream repository and pull the latest changes from the upstream repository.

kwon-dohun committed:
chore: revert unintended README modification

gunuzello committed:
refactor(settlement): define eps as class constant and remove unnecessary abs() call

- Moved the hardcoded epsilon (0.005) value into a class-level constant `eps` for better readability and maintainability.
- Removed the unnecessary `.abs()` call on c.getValue() since c always represents positive creditor balances.
- These changes improve code clarity, consistency, and make rounding tolerance easier to adjust later.

ohkyounghun committed:
Rollback .gitignore

ohkyounghun committed:
refactor: rename test method and add clarifying comments based on review feedback

- Renamed simpleTest() to test_singlePayment_whenOnePersonOwesAnother()
  to follow the 'test_whenCondition_thenExpectedResult' naming pattern.
- Added comments inside the loop to explain:
  • why null balances are skipped
  • how eps is used to filter tiny amounts
- Improves code readability and maintainability.

AyeongKwon commented on src/test/java/itm/oss/splitter/SettlementTest.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Test
+    void tinyAmountTest() {
```
Thanks for your great work! I have a small suggestion on your code. The current name tinyAmountTest is a bit vague what is the purpose of this function.

I suggest renaming it to:
```suggestion
    void test_ignoresPayments_whenBalanceIsBelowThreshold() {
```

AyeongKwon reviewed (commented):
Thanks for your quick response. 
Additionally, I recommend renaming this test function to clarify the purpose of the test.

Kimhyewon0621 reviewed (approved):
Thank you for applying my opinion! Everythings look very good~

AyeongKwon reviewed (approved):
LGTM!!

ohkyounghun committed:
quick fix

ohkyounghun committed:
Deleting DS_Store

ohkyounghun committed:
Update pom.xml

change java version 21 -> 17

ohkyounghun committed:
Update src/test/java/itm/oss/splitter/SettlementTest.java

Co-authored-by: AyeongKwon <zero523@seoultech.ac.kr>

ohkyounghun committed:
replace READ_ME into origin

ohkyounghun committed:
Merge branch 'feat/issue5-settlement' of https://github.com/ohkyounghun/expense-splitter into feat/issue5-settlement
